File: Python/Algorithm/Algorithm II/11. Container With Most Water.py
from typing import List


class Solution:
    def maxArea(self, height: List[int]) -> int:
        """
        Find the maximum area between two bars in a container.

        Args:
            heightList (List[int]): A list of integers representing the heights of the bars.

        Returns:
            int: The maximum area between two bars in the container.
        """
        HEIGHT_LENGTH = len(height)

        maxContainer = 0

        leftPointer = 0
        rightPointer = HEIGHT_LENGTH - 1

        while leftPointer < rightPointer:
            leftBar = height[leftPointer]
            rightBar = height[rightPointer]
            barDistance = rightPointer - leftPointer

            currentContainer = min(leftBar, rightBar) * barDistance
            maxContainer = max(maxContainer, currentContainer)

            if leftBar < rightBar:
                leftPointer += 1
            else:
                rightPointer -= 1

        return maxContainer

File: C/Algorithm/Algorithm II/11. Container With Most Water.c
/* Definition for getting the minimum value */
#define MIN(X, Y) (X < Y ? X : Y)

/* Definition for getting the maximum value */
#define MAX(X, Y) (X > Y ? X : Y)

/*
 * Calculates the maximum area between vertical bars.
 *
 * The 'maxArea' function takes an array 'height' representing the heights of vertical bars
 * and its size 'heightSize' as input. It calculates the maximum area between two vertical bars
 * by using a two-pointer approach. The function iterates through the array while adjusting the
 * pointers to maximize the area between the bars. The result is the maximum area that can be
 * enclosed by two bars.
 *
 * Parameters:
 * - height: An array representing the heights of vertical bars.
 * - heightSize: The number of elements in the 'height' array.
 *
 * Returns:
 * The maximum area between two vertical bars.
 */
int maxArea(int *height, int heightSize)
{
    int HEIGHT_LENGTH = heightSize;
    int maxContainer = 0;
    int leftPointer = 0;
    int rightPointer = HEIGHT_LENGTH - 1;

    while (leftPointer < rightPointer)
    {
        int leftBar = height[leftPointer];
        int rightBar = height[rightPointer];
        int barDistance = rightPointer - leftPointer;

        int currentContainer = MIN(leftBar, rightBar) * barDistance;
        maxContainer = MAX(maxContainer, currentContainer);

        if (leftBar < rightBar)
            leftPointer++;

        else
            rightPointer--;
    }

    return maxContainer;
}
